### PR TITLE
Add CodonFM (Encodon) codon-level masked-LM toolkit

### PR DESCRIPTION
**Pull request by roknabadi**
## Summary

Adds NVIDIA's **CodonFM / Encodon** — a codon-level, bidirectional (masked) language model over coding sequences — to proto as five tools under `masked_models/codonfm/`. Encodon reads a CDS as a sequence of codons and gives a learned, zero-shot view of how "natural" each codon and each coding sequence is, which is useful for codon optimization, variant effect ranking, and sequence design.

## The tools

| Tool | Returns | Use |
|---|---|---|
| `codonfm-fitness` | mean codon log-likelihood per sequence | rank whole CDS by naturalness |
| `codonfm-score` | ref-vs-alt log-likelihood ratio per mutation | score synonymous/missense codon substitutions |
| `codonfm-embeddings` | CLS embedding per sequence | featurize a CDS |
| `codonfm-gradient` | masked pseudo-log-likelihood gradient over relaxed `(L, 64)` codon logits | gradient-guided design (Fast SeqProp-style) |
| `codonfm-sample` | masked-codon resampling of a CDS | propose codon variants |

All follow the standard proto shape (`run_x(Input(...), Config(...))`), are GPU tools, and share `CodonFMConfig` (`model_checkpoint`, `device`, `batch_size`).

## Usage

```python
from proto_tools import (
    CodonFMFitnessInput, CodonFMFitnessConfig, run_codonfm_fitness,
    CodonFMScoreInput, CodonFMScoreConfig, run_codonfm_score,
)

# whole-CDS naturalness
fit = run_codonfm_fitness(
    CodonFMFitnessInput(sequences=[cds]),
    CodonFMFitnessConfig(model_checkpoint="encodon_80m"),
)
print(fit.results[0].fitness)          # mean codon log-likelihood

# score a synonymous substitution (codon_position is 1-based)
sc = run_codonfm_score(
    CodonFMScoreInput(mutations=[
        {"sequence": cds, "codon_position": 2, "ref_codon": "GCC", "alt_codon": "GCT"},
    ]),
    CodonFMScoreConfig(model_checkpoint="encodon_80m"),
)
print(sc.results[0].llr)               # + favors the reference codon
```

## How it works

- **Vendored source.** The Encodon model/tokenizer/inference code is vendored **byte-verbatim** under `standalone/src/` from `NVIDIA-BioNeMo/CodonFM` @ `0561907ca25adae7ecbe685e8fc2882e634b1f91` (provenance recorded in `standalone/src/VENDORED.md`), so the numerics match upstream exactly.
- **Standalone env.** Each tool runs in its own isolated venv built by `standalone/setup.sh` (pinned `torch==2.5.1`), driven through `standalone/inference.py` — no new dependencies added to the main package.
- **Checkpoints on demand.** `model_checkpoint` accepts `encodon_80m` / `encodon_600m` / `encodon_1b` / `encodon_1b_cdwt`, resolved via `CODONFM_CHECKPOINTS` and downloaded on first use from the **public** NVIDIA HuggingFace repos (e.g. `encodon_80m` → `nvidia/NV-CodonFM-Encodon-80M-v1`); **no HF token required**.
- **Vocabulary.** `CODONFM_CODON_VOCAB` is the 64 DNA codons (`product("ACGT", 3)`, including the three stop codons) in the tokenizer's column order; the 5 special tokens are excluded.

## Example notebook

`examples/example.ipynb` codon-optimizes a **real gene** (human preproinsulin, NCBI NM_000207) end to end: it tours all five tools, measures the classic design metrics (CAI, MFE via ViennaRNA, GC%, U%), shows that **Encodon's learned fitness already correlates with CAI** (so it is not an independent objective), and runs an iterative, context-re-scored synonymous optimization that pulls CAI up and MFE down together while preserving the protein.

## Tests

`tests/masked_models_tests/test_codonfm.py` covers the input/config validation, the 1-based↔0-based `codon_position` handling, payload/response contracts for each tool, and the sense-codon sampling guard.

## Notes for reviewers

- **`codonfm-sample` draws only from the 61 sense codons**, so a resampled interior codon can never become a premature stop; the tokenizer vocab itself still includes the 3 stops (documented in `shared_data_models.py`).
- **`codonfm-gradient` normalizes its cross-entropy over the full 69-token vocabulary**, matching upstream `predict_fitness` / `predict_mutation`; the codon-restricted softmax is a sampling-only choice in `codonfm-sample`.
- **`codonfm-fitness` is the mean over non-padding tokens**, which per the upstream definition includes the CLS and SEP tokens — so it is a per-token, not strictly per-codon, mean. Noted in the fitness docstring.
- **`codonfm-score` uses 1-based `codon_position`** at the tool boundary and converts to 0-based before dispatch.
- The standalone env build copies helpers with a **content-only copy** for GCSFuse compatibility (`_worker_bootstrap.py`), avoiding the `copystat` EPERM that breaks `shutil.copy2`/`copytree` on mounted buckets.

## License

Encodon code is Apache-2.0; weights are under the NVIDIA Open Model License. Both are recorded in `license.yaml`; upstream is cited in `cite.bib`.

### CODE REVIEW

**Comment by brianhie** [comment]
I resolved the merge conflict against current `main` and completed the CodonFM review on [`review/pr-34-codonfm-ready`](https://github.com/evo-design/proto-tools/tree/review/pr-34-codonfm-ready).

Changes include current iterable/config/device-worker contracts, a per-item sampling result model, public-weight/license corrections, exact vendored-source exclusions/provenance, a production Modal service/tool-map entry for all five tools, and a fully executed and sanitized example notebook with the missing sampling example and hero image.

Verified locally with an H100 GPU.

**Comment by brianhie** [comment]
@roknabadi I made a few changes on top of your PR.

A compact way of viewing these is here: https://github.com/thernabio/proto-tools/compare/add-codonfm-tools...evo-design:proto-tools:review/pr-34-codonfm-only

The full set of changes includes a big update to the latest evo-design/main structure and has been turned into a PR here: https://github.com/thernabio/proto-tools/pull/1

Happy to change in any way.

**Comment by roknabadi** [comment]
Hi @brianhie, the changes look good to me. Thank you so much for resolving the conflicts. Appreciate the help!

**Comment by brianhie** [comment]
@bviggiano can you also take a look? This PR lgtm.

**Comment by bviggiano** [comment]
Looks great! Making a few changes so this can use our standardized masking strategy infra. I'll have this PR merged today

**Comment by bviggiano** [comment]
@roknabadi Folded in a few more changes on https://github.com/thernabio/proto-tools/pull/2! Feel free to make any updates on this if you feel they are needed, then this should be good to merge!

**Comment by bviggiano** [comment]
@roknabadi Sorry one more to merge when you have a moment to fix the CI check above! I can't push to this branch directly for some reason 🤷‍♂️ https://github.com/thernabio/proto-tools/pull/3

**Review by bviggiano — APPROVED** [review]
